File: ocaml/binary-search/binary_search.ml
let rec search array value = function
  | left, right when left > right -> Error "value not in array"
  | left, right ->
    let mid = (left + right) / 2 in
    match array.(mid) with
    | v when v = value -> Ok mid
    | v when v < value -> search array value (mid + 1, right)
    | _ -> search array value (left, mid - 1)

let find array value =
    search array value (0, Array.length array - 1)

File: fortran/binary-search/binary_search.f90
module binary_search
  implicit none
contains

  function find(array, val) result(idx)
    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx, left, right, mid

    left = 1
    right = size(array)
    idx = -1

    do while (left <= right)
      mid = (left + right) / 2
      if (array(mid) == val) then
        idx = mid
        exit
      else if (array(mid) < val) then
        left = mid + 1
      else
        right = mid - 1
      end if
    end do
  end function

end module
